### PR TITLE
FOUNDATIONS: Modify Group Posts

### LINKED ISSUE #390: FOUNDATIONS: Modify Group Posts

### PR DESCRIPTION
**Pull request by jasurbekyusuf**
Closes #390

### CODE REVIEW

**Review by GulchexrasGithub — APPROVED** [review]

File: Taarafo.Core/Models/GroupPosts/GroupPost.cs [hunk 1]
```diff
@@ -16,5 +16,7 @@ public class GroupPost
 
 		public Guid PostId { get; set; }
 		public Post Post { get; set; }
-	}
+        public DateTimeOffset CreatedDate { get; set; }
```
Comment (cjdutoit) [inline Taarafo.Core/Models/GroupPosts/GroupPost.cs:19]: Remove this line, we don't need this here

File: Taarafo.Core/Models/GroupPosts/GroupPost.cs [hunk 2]
```diff
@@ -16,5 +16,7 @@ public class GroupPost
...
+        public DateTimeOffset UpdatedDate { get; set; }
```
Comment (cjdutoit) [inline Taarafo.Core/Models/GroupPosts/GroupPost.cs:20]: Remove this line, we don't need this here

File: Taarafo.Core/Services/Foundations/GroupPosts/GroupPostService.cs [hunk 3]
```diff
@@ -53,6 +53,20 @@ public partial class GroupPostService : IGroupPostService
         public IQueryable<GroupPost> RetrieveAllGroupPosts() =>
             TryCatch(() => this.storageBroker.SelectAllGroupPosts());
 
+        public ValueTask<GroupPost> ModifyGroupPostAsync(GroupPost groupPost) =>
```
Comment (cjdutoit) [inline Taarafo.Core/Services/Foundations/GroupPosts/GroupPostService.cs:56]: Remove the DateTimeBroker as we do not need this dependency

File: Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.cs [hunk 4]
```diff
@@ -38,6 +39,21 @@ public GroupPostServiceTests()
                 loggingBroker: this.loggingBrokerMock.Object);
         }
 
+        public static TheoryData MinutesBeforeOrAfter()
```
Comment (cjdutoit) [inline Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.cs:42]: Remove the DateTimeBroker as we do not need this dependency

File: Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+            var invalidGroupPostException = new InvalidGroupPostException();
+
+            invalidGroupPostException.AddData(
+                key: nameof(GroupPost.GroupId),
+                values: "Id is required");
+
+            invalidGroupPostException.AddData(
+                key: nameof(GroupPost.PostId),
+                values: "Id is required");
+
```
Comment (cjdutoit) [inline Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs:71]: Remove extra line

File: Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+                        "Value is required",
+                        "Date is not recent",
+                        $"Date is the same as {nameof(GroupPost.CreatedDate)}"
+                    }
+            );
+
+            var expectedGroupPostValidationException =
+                new GroupPostValidationException(invalidGroupPostException);
+
+            this.dateTimeBrokerMock.Setup(broker =>
```
Comment (cjdutoit) [inline Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs:90]: Remove this

File: Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+
+            GroupPostValidationException actualGroupPostValidationException =
+                await Assert.ThrowsAsync<GroupPostValidationException>(
+                    modifyGroupPostTask.AsTask);
+
+            //then
+            actualGroupPostValidationException.Should()
+                .BeEquivalentTo(expectedGroupPostValidationException);
+
+            this.dateTimeBrokerMock.Verify(broker =>
```
Comment (cjdutoit) [inline Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs:106]: Remove this

File: Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+                broker.GetCurrentDateTimeOffset(), Times.Once);
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogError(It.Is(SameExceptionAs(
+                    expectedGroupPostValidationException))), Times.Once);
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.UpdateGroupPostAsync(It.IsAny<GroupPost>()), Times.Never);
+
+            this.dateTimeBrokerMock.VerifyNoOtherCalls();
```
Comment (cjdutoit) [inline Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs:116]: remove this

File: Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+            this.storageBrokerMock.VerifyNoOtherCalls();
+        }
+
+        [Fact]
+        public async Task ShouldThrowValidationExceptionOnModifyIfUpdatedDateIsSameAsCreatedDateAndLogItAsync()
+        {
+            // given
+            DateTimeOffset randomDateTime = GetRandomDateTimeOffset();
```
Comment (cjdutoit) [inline Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs:125]: Remove this

File: Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+            GroupPost randomGroupPost = CreateRandomGroupPost(randomDateTime);
```
Comment (cjdutoit) [inline Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs:126]: change to `GroupPost randomGroupPost = CreateRandomGroupPost();`

File: Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs [hunk 11]
```diff
@@ -0,0 +1,370 @@
...
+            this.storageBrokerMock.Verify(broker =>
+                broker.UpdateGroupPostAsync(It.IsAny<GroupPost>()), Times.Never);
+
+            this.dateTimeBrokerMock.VerifyNoOtherCalls();
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+            this.storageBrokerMock.VerifyNoOtherCalls();
+        }
+
+        [Fact]
+        public async Task ShouldThrowValidationExceptionOnModifyIfUpdatedDateIsSameAsCreatedDateAndLogItAsync()
```
Comment (cjdutoit) [inline Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs:122]: Can remove this test as we do not have any dates on GroupPost

File: Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs [hunk 12]
```diff
@@ -0,0 +1,370 @@
...
+                    invalidGroupPost.PostId), Times.Never);
+
+            this.dateTimeBrokerMock.VerifyNoOtherCalls();
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+            this.storageBrokerMock.VerifyNoOtherCalls();
+        }
+
+        [Theory]
+        [MemberData(nameof(MinutesBeforeOrAfter))]
+        public async Task ShouldThrowValidationExceptionOnModifyIfUpdatedDateIsNotRecentAndLogItAsync(int minutes)
```
Comment (cjdutoit) [inline Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs:170]: Can remove this test as we do not have any dates on GroupPost

File: Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs [hunk 13]
```diff
@@ -0,0 +1,370 @@
...
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+            this.storageBrokerMock.VerifyNoOtherCalls();
+        }
+
+        [Fact]
+        public async Task ShouldThrowValidationExceptionOnModifyIfGroupPostDoesNotExistAndLogItAsync()
+        {
+            // given
+            int randomNegativeMinutes = GetRandomNegativeNumber();
+            DateTimeOffset dateTime = GetRandomDateTimeOffset();
```
Comment (cjdutoit) [inline Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs:220]: Update this as above...remove the DateTime

File: Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs [hunk 14]
```diff
@@ -0,0 +1,370 @@
...
+                await Assert.ThrowsAsync<GroupPostValidationException>(modifyGroupPostTask.AsTask);
+
+            // then
+            actualGroupPostValidationException.Should().BeEquivalentTo(expectedGroupPostValidationException);
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.SelectGroupPostByIdAsync(nonExistGroupPost.GroupId,
+                    nonExistGroupPost.PostId), Times.Once);
+
+            this.dateTimeBrokerMock.Verify(broker =>
```
Comment (cjdutoit) [inline Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs:254]: Remove this

File: Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs [hunk 15]
```diff
@@ -0,0 +1,370 @@
...
+                broker.GetCurrentDateTimeOffset(), Times.Once);
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogError(It.Is(SameExceptionAs(
+                    expectedGroupPostValidationException))), Times.Once);
+
+            this.storageBrokerMock.VerifyNoOtherCalls();
+            this.dateTimeBrokerMock.VerifyNoOtherCalls();
```
Comment (cjdutoit) [inline Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs:262]: remove this

File: Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs [hunk 16]
```diff
@@ -0,0 +1,370 @@
...
+        }
+
+        [Fact]
+        public async Task ShouldThrowValidationExceptionOnModifyIfStorageCreatedDateNotSameAsCreatedDateAndLogItAsync()
+        {
+            // given
+            int randomNumber = GetRandomNegativeNumber();
+            int randomMinutes = randomNumber;
+            DateTimeOffset randomDateTime = GetRandomDateTimeOffset();
+            GroupPost randomGroupPost = CreateRandomModifyGroupPost(randomDateTime);
```
Comment (cjdutoit) [inline Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs:273]: Update as above

File: Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs [hunk 17]
```diff
@@ -0,0 +1,370 @@
...
+                key: nameof(GroupPost.CreatedDate),
+                values: $"Date is not same as {nameof(GroupPost.CreatedDate)}");
+
+            var expectedGroupPostValidationException =
+                new GroupPostValidationException(invalidGroupPostException);
+
+            this.storageBrokerMock.Setup(broker =>
+                broker.SelectGroupPostByIdAsync(GroupId, PostId)).ReturnsAsync(storageGroupPost);
+
+            this.dateTimeBrokerMock.Setup(broker =>
```
Comment (cjdutoit) [inline Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs:292]: Remove this

File: Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs [hunk 18]
```diff
@@ -0,0 +1,370 @@
...
+                await Assert.ThrowsAsync<GroupPostValidationException>(modifyGroupPostTask.AsTask);
+
+            // then
+            actualGroupPostValidationException.Should().BeEquivalentTo(expectedGroupPostValidationException);
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.SelectGroupPostByIdAsync(invalidGroupPost.GroupId,
+                    invalidGroupPost.PostId), Times.Once);
+
+            this.dateTimeBrokerMock.Verify(broker =>
```
Comment (cjdutoit) [inline Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs:308]: Remove this

File: Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs [hunk 19]
```diff
@@ -0,0 +1,370 @@
...
+                values: $"Date is the same as {nameof(GroupPost.UpdatedDate)}");
+
+            var expectedGroupPostValidationException =
+                new GroupPostValidationException(invalidGroupPostException);
+
+            this.storageBrokerMock.Setup(broker =>
+               broker.SelectGroupPostByIdAsync(invalidGroupPost.GroupId,
+                   invalidGroupPost.PostId)).ReturnsAsync(storageGroupPost);
+
+            this.dateTimeBrokerMock.Setup(broker =>
```
Comment (cjdutoit) [inline Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs:344]: Remove

File: Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs [hunk 20]
```diff
@@ -0,0 +1,370 @@
...
+            GroupPostValidationException actualGroupPostValidationException =
+               await Assert.ThrowsAsync<GroupPostValidationException>(modifyGroupPostTask.AsTask);
+
+            // then
+            actualGroupPostValidationException.Should().BeEquivalentTo(expectedGroupPostValidationException);
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.SelectGroupPostByIdAsync(groupId, postId), Times.Once);
+
+            this.dateTimeBrokerMock.Verify(broker =>
```
Comment (cjdutoit) [inline Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Modify.cs:358]: remove

File: Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Exceptions.Modify.cs [hunk 21]
```diff
@@ -0,0 +1,226 @@
...
+namespace Taarafo.Core.Tests.Unit.Services.Foundations.GroupPosts
+{
+    public partial class GroupPostServiceTests
+    {
+        [Fact]
+        public async Task ShouldThrowCriticalDependencyExceptionOnModifyIfSqlErrorOccursAndLogItAsync()
+        {
+            // given
+            DateTimeOffset someDateTime = GetRandomDateTimeOffset();
+            GroupPost randomGroupPost = CreateRandomGroupPost(someDateTime);
```
Comment (cjdutoit) [inline Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Exceptions.Modify.cs:25]: Remove the datetime

File: Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Exceptions.Modify.cs [hunk 22]
```diff
@@ -0,0 +1,226 @@
...
+            GroupPostDependencyException actualGroupPostDependencyException =
+                await Assert.ThrowsAsync<GroupPostDependencyException>(
+                     modifyGroupPostTask.AsTask);
+
+            // then
+            actualGroupPostDependencyException.Should().BeEquivalentTo(
+                expectedGroupPostDependencyException);
+
+            this.dateTimeBrokerMock.Verify(broker =>
+                broker.GetCurrentDateTimeOffset(), Times.Once);
```
Comment (cjdutoit) [inline Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Exceptions.Modify.cs:53]: change this to:
```
    this.dateTimeBrokerMock.Verify(broker =>
        broker.GetCurrentDateTimeOffset(), Times.Once);
```
also update the other places

**Review by cjdutoit — CHANGES_REQUESTED** [review]
Some changes required

**Review by cjdutoit — APPROVED** [review]

**Pull request merged**